File: dotnet/Allmark/Render/FootnoteRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class FootnoteRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "footnote",
            Render = Render,
        };
    }

    public static void Render(MarkdownNode node, RendererState state, bool? decode = true)
    {
        if (!state.Footnotes.Any((f) => f.Info == node.Info))
        {
            state.Footnotes.Add(node);
        }
        var label = state.Footnotes.Count;
        var id = $"fnref{label}";
        var href = $"#fn{label}";
        state.Output.Append($"<sup class=\"footnote-ref\"><a href=\"{href}\" id=\"{id}\">{label}</a></sup>");
    }
}

File: web/src/render-console/footnoteRenderer.ts
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";
import ANSI from "./ansi";

const renderer: Renderer = {
	name: "footnote",
	render,
};
export default renderer;

function render(node: MarkdownNode, state: RendererState): void {
	const style = ANSI.dim;
	const reset = ANSI.reset;
	if (state.footnotes.find((f) => f.info === node.info) === undefined) {
		state.footnotes.push(node);
	}
	const label = state.footnotes.length;
	state.output += `${style}[${label}]${reset}`;
}
